### PR TITLE
[FLINK-2790] [yarn] [ha] Adds high availability support for Yarn

### PR DESCRIPTION
**Pull request by tillrohrmann**
Adds high availability support for Yarn by exploiting Yarn's functionality to restart a failed application master. Depending on the Hadoop version the behaviour is an increasing superset of functionalities of the preceding version's behaviour

### 2.3.0 <= version < 2.4.0

* Set the number of application attempts to the configuration value `yarn.application-attempts`. This means that the application can be restarted `yarn.application-attempts` time before yarn fails the application. In case of an application master, all other task manager containers will also be killed. It's important to note that the number of application attempts set within Flink cannot exceed the yarn cluster setting `yarn.resourcemanager.am.max-attempts` with which the Yarn cluster was started. Thus, `yarn.resourcemanager.am.max-attempts` is the upper bound for application restarts.

### 2.4.0 <= version < 2.6.0

* Additionally, enables that containers will be kept across application attempts. This avoids the killing of TaskManager containers in the case of an application master failure. This has the advantage that the startup time is faster and that the user does not have to wait for obtaining the container resources again.

### 2.6.0 <= version

* Sets the attempts failure validity interval to the akka timeout value. The attempts failure validity interval says that an application is only killed after the system has seen the maximum number of application attempts during one interval. This avoids that a long lasting job will deplete it's application attempts.

This PR also refactors the different Yarn components to allow the start-up of testing actors within Yarn. Furthermore, the `JobManager` start up logic is slightly extended to allow code reuse in the `ApplicationMasterBase`.

The HA functionality is tested via the `YARNHighAvailabilityITCase` where an application master is multiple times killed. Each time it's checked that the single TaskManager successfully reconnects to the newly started `YarnJobManager`. In case of version `2.3.0`, the `TaskManager` is restarted.

### CODE REVIEW

**Comment by rmetzger** [comment]
I've looked a bit over the code and it all looks like good changes.

I've tested the code on a Hadoop 2.6.0 cluster and the YARN session was restarting properly. But I think that's not the expected behavior.
We actually want the TMs to stay alive and reconnect to the new AM.

**Comment by tillrohrmann** [comment]
I think TMs are only kept alive if their containers have been properly started. If the AM happens to die while the TM container are started up, I think they will be terminated as well. Another question is how did you kill the AM and what do you mean with "[...] restarting properly. But I think that's not the expected behavior"?

**Comment by uce** [comment]
Looks like a lot of work to figure out the different version behaviours. Good job and thanks for the clear explanation. :)

I guess Robert meant with "not restarting properly" that the TMs were restarted as well.

How does the way you kill the AM affect recovery?

I will try this out later today.

**Comment by tillrohrmann** [comment]
I was just curious whether he killed them gracefully with a `PoisonPill` or via killing the JVM process.

**Comment by rmetzger** [comment]
I killed the AM using "kill" from the command line, and all the Taskmanagers were shutting down

**Comment by tillrohrmann** [comment]
I found the issue. My code did not properly reflected the methods `setKeepContainersAcrossApplicationAttempts` and `setAttemptFailuresValidityInterval`. With the fix, already started containers are retained. Tested it with Yarn 2.7.1.

**Comment by tillrohrmann** [comment]
Good catch @rmetzger :-)

File: flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingCluster.scala [hunk 1]
```diff
@@ -103,18 +103,17 @@ class TestingCluster(
...
       scheduler,
       libraryCacheManager,
-      _,
       executionRetries,
       delayBetweenRetries,
       timeout,
       archiveCount,
-      leaderElectionService) = JobManager.createJobManagerComponents(config)
+      leaderElectionService) = JobManager.createJobManagerComponents(
+      config,
```
Comment (uce) [inline flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingCluster.scala]: indentation

File: flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingJobManagerLike.scala [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+import org.apache.flink.runtime.testingUtils.TestingTaskManagerMessages.AccumulatorsChanged
+
+import scala.collection.mutable
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+import language.postfixOps
+
+/** This mixin can be used to decorate a JobManager with messages for testing purpose.
+  *
```
Comment (uce) [inline flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingJobManagerLike.scala]: empty line

File: flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingJobManagerLike.scala [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+              sender ! decorateMessage(
+                WorkingTaskManager(
+                  Some(resource.getInstance().getActorGateway)
+                )
+              )
+            }
+          }
+        case None => sender ! decorateMessage(WorkingTaskManager(None))
+      }
+
```
Comment (uce) [inline flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingJobManagerLike.scala]: empty line

File: flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingTaskManagerLike.scala [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+import org.apache.flink.runtime.testingUtils.TestingMessages.{DisableDisconnect,
+CheckIfJobRemoved, Alive}
+import org.apache.flink.runtime.testingUtils.TestingTaskManagerMessages._
+
+import scala.concurrent.duration._
+
+import language.postfixOps
+
+/** This mixin can be used to decorate a TaskManager with messages for testing purposes.
+  *
```
Comment (uce) [inline flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingTaskManagerLike.scala]: empty line

File: flink-yarn/src/main/java/org/apache/flink/yarn/FlinkYarnClientBase.java [hunk 5]
```diff
@@ -0,0 +1,867 @@
...
+			if(hasLogback) {
+				amCommand += " -Dlogback.configurationFile=file:" + FlinkYarnSessionCli.CONFIG_FILE_LOGBACK_NAME;
+			}
+
+			if(hasLog4j) {
+				amCommand += " -Dlog4j.configuration=file:" + FlinkYarnSessionCli.CONFIG_FILE_LOG4J_NAME;
+			}
+		}
+
+		amCommand 	+= " " + getApplicationMasterClass().getName() + " "
```
Comment (uce) [inline flink-yarn/src/main/java/org/apache/flink/yarn/FlinkYarnClientBase.java]: indentation

**Comment by uce** [comment]
The code looks good. :-) I like the changes around the testing classes and the changes to the retrieval utils (that one I've looked into in more detail). I didn't check the YARN logic in detail though. I'm going to try it out now locally and report back. I expect that to work just fine and then we can safely merge this I think. :)

**Comment by uce** [comment]
I have a dependency problem with Curator leading to:

```bash
ERROR org.apache.flink.runtime.jobmanager.JobManager                - Error while starting up JobManager
java.lang.NoSuchMethodError: org.apache.curator.utils.PathUtils.validatePath(Ljava/lang/String;)Ljava/lang/String;
    at org.apache.curator.framework.imps.NamespaceImpl.<init>(NamespaceImpl.java:37)
    at org.apache.curator.framework.imps.CuratorFrameworkImpl.<init>(CuratorFrameworkImpl.java:113)
    at org.apache.curator.framework.CuratorFrameworkFactory$Builder.build(CuratorFrameworkFactory.java:124)
    at org.apache.flink.runtime.util.ZooKeeperUtils.startCuratorFramework(ZooKeeperUtils.java:83)
    at org.apache.flink.runtime.util.ZooKeeperUtils.createLeaderElectionService(ZooKeeperUtils.java:145)
    at org.apache.flink.runtime.util.LeaderElectionUtils.createLeaderElectionService(LeaderElectionUtils.java:52)
    at org.apache.flink.runtime.jobmanager.JobManager$.createJobManagerComponents(JobManager.scala:1595)
    at org.apache.flink.runtime.jobmanager.JobManager$.startJobManagerActors(JobManager.scala:1672)
    at org.apache.flink.runtime.jobmanager.JobManager$.startJobManagerActors(JobManager.scala:1629)
    at org.apache.flink.runtime.jobmanager.JobManager$.startActorSystemAndJobManagerActors(JobManager.scala:1307)
    at org.apache.flink.yarn.ApplicationMasterBase.runAction(ApplicationMasterBase.scala:127)
    at org.apache.flink.yarn.ApplicationMasterBase$$anon$1.run(ApplicationMasterBase.scala:76)
    at java.security.AccessController.doPrivileged(Native Method)
    at javax.security.auth.Subject.doAs(Subject.java:360)
    at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1608)
    at org.apache.flink.yarn.ApplicationMasterBase.run(ApplicationMasterBase.scala:74)
    at org.apache.flink.yarn.ApplicationMaster$.main(ApplicationMaster.scala:35)
    at org.apache.flink.yarn.ApplicationMaster.main(ApplicationMaster.scala)
```

Our application master class path:
```bash
13:45:50,795 DEBUG org.apache.flink.yarn.ApplicationMaster                       - All environment variables: {PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin, HADOOP_CONF_DIR=/Users/ufuk/Downloads/hadoop-2.6.0/etc/hadoop, MAX_APP_ATTEMPTS=2, HADOOP_SECURE_DN_PID_DIR=, HADOOP_PID_DIR=, MAIL=/var/mail/ufuk, LD_LIBRARY_PATH=:/Users/ufuk/Downloads/hadoop-2.6.0/lib/native:/Users/ufuk/Downloads/hadoop-2.6.0/lib/native, LOGNAME=ufuk, JVM_PID=13307, _DETACHED=false, PWD=/tmp/hadoop-ufuk/nm-local-dir/usercache/ufuk/appcache/application_1444217271951_0004/container_1444217271951_0004_01_000001, HADOOP_YARN_USER=yarn, HADOOP_PREFIX=/Users/ufuk/Downloads/hadoop-2.6.0, LOCAL_DIRS=/tmp/hadoop-ufuk/nm-local-dir/usercache/ufuk/appcache/application_1444217271951_0004, YARN_IDENT_STRING=ufuk, HADOOP_SECURE_DN_LOG_DIR=/, SHELL=/bin/zsh, YARN_CONF_DIR=/Users/ufuk/Downloads/hadoop-2.6.0/etc/hadoop, JAVA_MAIN_CLASS_10305=org.apache.hadoop.yarn.server.nodemanager.NodeManager, LOG_DIRS=/Users/ufuk/Downloads/hadoop-2.6.0/logs/userlogs/application_1444217271951_0004/container_1444217271951_0004_01_000001, _CLIENT_SHIP_FILES=file:/Users/ufuk/.flink/application_1444217271951_0004/flink-python-0.10-SNAPSHOT.jar,file:/Users/ufuk/.flink/application_1444217271951_0004/log4j-1.2.17.jar,file:/Users/ufuk/.flink/application_1444217271951_0004/slf4j-log4j12-1.7.7.jar,file:/Users/ufuk/.flink/application_1444217271951_0004/logback.xml,file:/Users/ufuk/.flink/application_1444217271951_0004/log4j.properties, _CLIENT_USERNAME=ufuk, HADOOP_YARN_HOME=/Users/ufuk/Downloads/hadoop-2.6.0, TMPDIR=/var/folders/_c/5tc5q5q55qjcjtqwlwvwd1m00000gn/T/, HADOOP_DATANODE_OPTS=-Dhadoop.security.logger=ERROR,RFAS -Dhadoop.security.logger=ERROR,RFAS , HADOOP_SECONDARYNAMENODE_OPTS=-Dhadoop.security.logger=INFO,RFAS -Dhdfs.audit.logger=INFO,NullAppender -Dhadoop.security.logger=INFO,RFAS -Dhdfs.audit.logger=INFO,NullAppender , _FLINK_JAR_PATH=file:/Users/ufuk/.flink/application_1444217271951_0004/flink-dist-0.10-SNAPSHOT.jar, __CF_USER_TEXT_ENCODING=0x1F5:0x0:0x0, LC_CTYPE=UTF-8, _CLIENT_TM_COUNT=2, _CLIENT_TM_MEMORY=1024, SHLVL=3, HADOOP_IDENT_STRING=ufuk, YARN_ROOT_LOGGER=INFO,RFA, _SLOTS=-1, _CLIENT_HOME_DIR=file:/Users/ufuk, APP_SUBMIT_TIME_ENV=1444218347305, NM_HOST=192.168.178.69, _APP_ID=application_1444217271951_0004, YARN_LOGFILE=yarn-ufuk-nodemanager-vinci.local.log, HADOOP_SECURE_DN_USER=, HADOOP_CLASSPATH=/contrib/capacity-scheduler/*.jar:/contrib/capacity-scheduler/*.jar, HADOOP_HDFS_HOME=/Users/ufuk/Downloads/hadoop-2.6.0, HADOOP_MAPRED_HOME=/Users/ufuk/Downloads/hadoop-2.6.0, HADOOP_COMMON_HOME=/Users/ufuk/Downloads/hadoop-2.6.0, HADOOP_CLIENT_OPTS=-Xmx512m -Xmx512m , _=/bin/java, APPLICATION_WEB_PROXY_BASE=/proxy/application_1444217271951_0004, _STREAMING_MODE=false, NM_HTTP_PORT=8042, JAVA_MAIN_CLASS_13316=org.apache.flink.yarn.ApplicationMaster, HADOOP_OPTS= -Djava.net.preferIPv4Stack=true -Dhadoop.log.dir=/Users/ufuk/Downloads/hadoop-2.6.0/logs -Dhadoop.log.file=hadoop.log -Dhadoop.home.dir=/Users/ufuk/Downloads/hadoop-2.6.0 -Dhadoop.id.str=ufuk -Dhadoop.root.logger=INFO,console -Djava.library.path=/Users/ufuk/Downloads/hadoop-2.6.0/lib/native -Dhadoop.policy.file=hadoop-policy.xml -Djava.net.preferIPv4Stack=true -Djava.net.preferIPv4Stack=true -Dhadoop.log.dir=/Users/ufuk/Downloads/hadoop-2.6.0/logs -Dhadoop.log.file=hadoop.log -Dhadoop.home.dir=/Users/ufuk/Downloads/hadoop-2.6.0 -Dhadoop.id.str=ufuk -Dhadoop.root.logger=INFO,console -Djava.library.path=/Users/ufuk/Downloads/hadoop-2.6.0/lib/native -Dhadoop.policy.file=hadoop-policy.xml -Djava.net.preferIPv4Stack=true, SSH_CLIENT=::1 60015 22, NM_PORT=60017, USER=ufuk, CLASSPATH=/tmp/hadoop-ufuk/nm-local-dir/usercache/ufuk/appcache/application_1444217271951_0004/container_1444217271951_0004_01_000001/*:/Users/ufuk/Downloads/hadoop-2.6.0/etc/hadoop:/Users/ufuk/Downloads/hadoop-2.6.0/share/hadoop/common/*:/Users/ufuk/Downloads/hadoop-2.6.0/share/hadoop/common/lib/*:/Users/ufuk/Downloads/hadoop-2.6.0/share/hadoop/hdfs/*:/Users/ufuk/Downloads/hadoop-2.6.0/share/hadoop/hdfs/lib/*:/Users/ufuk/Downloads/hadoop-2.6.0/share/hadoop/yarn/*:/Users/ufuk/Downloads/hadoop-2.6.0/share/hadoop/yarn/lib/*, SSH_CONNECTION=::1 60015 ::1 22, HADOOP_TOKEN_FILE_LOCATION=/tmp/hadoop-ufuk/nm-local-dir/usercache/ufuk/appcache/application_1444217271951_0004/container_1444217271951_0004_01_000001/container_tokens, HADOOP_NFS3_OPTS=, HADOOP_NAMENODE_OPTS=-Dhadoop.security.logger=INFO,RFAS -Dhdfs.audit.logger=INFO,NullAppender -Dhadoop.security.logger=INFO,RFAS -Dhdfs.audit.logger=INFO,NullAppender , YARN_NICENESS=0, HOME=/home/, CONTAINER_ID=container_1444217271951_0004_01_000001, HADOOP_PORTMAP_OPTS=-Xmx512m -Xmx512m , MALLOC_ARENA_MAX=4}
```

The problem is */Users/ufuk/Downloads/hadoop-2.6.0/share/hadoop/yarn/lib/**, which contains Curator 2.6.

I've used vanilla YARN 2.6.0. Robert for whom it worked used cdh 5.4.

File: flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingCluster.scala (hunk 1 above)
Comment (tillrohrmann) [inline flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingCluster.scala]: Good catch

File: flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingJobManagerLike.scala (hunk 2 above)
Comment (tillrohrmann) [inline flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingJobManagerLike.scala]: Let's see whether it still recognized as a ScalaDoc

File: flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingJobManagerLike.scala (hunk 3 above)
Comment (tillrohrmann) [inline flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingJobManagerLike.scala]: fixed

File: flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingTaskManagerLike.scala (hunk 4 above)
Comment (tillrohrmann) [inline flink-runtime/src/test/scala/org/apache/flink/runtime/testingUtils/TestingTaskManagerLike.scala]: Fixed

File: flink-yarn/src/main/java/org/apache/flink/yarn/FlinkYarnClientBase.java (hunk 5 above)
Comment (tillrohrmann) [inline flink-yarn/src/main/java/org/apache/flink/yarn/FlinkYarnClientBase.java]: Fixed

**Comment by uce** [comment]
Smooth experience now. Works as expected with vanilla YARN 2.6.0.

+1 to merge.

**Comment by tillrohrmann** [comment]
Thanks for the review @uce. I addressed your comments.

I fixed the problem with Hadoop 2.6.0 by relocating Flink's curator dependency to `org.apache.flink.shaded.org.apache.curator`. Furthermore, I bumped `flink-shaded-curator's` Guava version to Flink's Guava version so that we don't include too many different Guava versions in the resulting `flink-dist.jar`.

**Pull request closed without merging**